typeclass-experiments/src/Common/Types.hs

```diff
@@ -0,0 +1,175 @@
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
+
+module Common.Types where
+
+import Common.SourceTy qualified as Ty
+import Data.Bifunctor
+import Data.List
+import Data.List.NonEmpty (NonEmpty (..))
+import Data.List.NonEmpty qualified as NE
+import Data.Set (Set)
+import Data.Text (Text)
+
+{- A simple ADT to represent types (i.e. the terms of our schema language).
+
+We can also use this to represent patterns by using PatVars in compound types.
+
+Note that this ADT allows us to represent nonsensical types. I wanted to prove to Drazen that I could
+do something without using a billion language extensions, and some nonsense is the price we pay for that :p
+-}
+
+data Pat
```

Comment:
Yeah I still don't understand why you're reformulating everything into Pat? It looks isomorphic.

What are you actually trying to do? Have a query on types or? Add some documentation and examples, because it doesn't click for me.

If you're trying to do querying on types you can definitely use https://hackage.haskell.org/package/unification-fd-0.11.2/docs/Control-Unification.html and represent our types with UTerm and Variable. You get the ability to 'pattern match' ie. do basic unification query on pairs of type terms. Is that at all what you're trying to do?

Reply:
Because it *is* a pattern, and not a type. It can have holes for variables in places where you can't put a type variable. It becomes a type when it gets instantiated such that the only variables in the pattern are in places where type variables can legitimately live, or when there aren't any variables in the pattern.

Reply:
When I say "it is a pattern", I mean it's analogous to [Pat in the TH AST](https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH-Syntax.html#t:Pat). Which is similar, but not identical to [Exp in the TH AST](https://hackage.haskell.org/package/template-haskell-2.19.0.0/docs/Language-Haskell-TH-Syntax.html#t:Exp).

Reply:
Ok thanks! I suspected you had something principled like this in mind, let me dig through the links you sent out and go through the whole thing again.

Reply:
So in TH Pat just represents 'pattern expressions' afaict, as in

```haskell
\Just (x:xs) -> ...
\pppppppp -> eee
```

Pat is `p` right?